subsys/nrf_rpc/nrf_rpc_uart.c

```diff
@@ -0,0 +1,302 @@
+ */
+void serial_cb(const struct device *uart, void *user_data)
+{
+	const struct nrf_rpc_tr *transport = (struct nrf_rpc_tr *)user_data;
+	struct nrf_rpc_uart *uart_config = transport->ctx;
+
+	if (!uart_irq_update(uart)) {
+		return;
+	}
+
+	while (uart_irq_rx_ready(uart)) {
+		uint8_t byte = 0; /* Have to assign to stop GCC from whining */
+
+		uart_fifo_read(uart, &byte, 1);
+		uint32_t ret = ring_buf_put(uart_config->ringbuf, &byte, 1);
+		(void)ret; /* Don't warn if logging is disabled */
+
+		LOG_DBG("rx: %x, rb put %u", byte, ret);
+
+		/* Only try to decode if wq item is not pending */
+		if (!k_work_busy_get(&uart_config->work)) {
+			process_ringbuf(uart_config);
+		}
+	}
+}
+
+static int init(const struct nrf_rpc_tr *transport,
+		nrf_rpc_tr_receive_handler_t receive_cb,
+		void *context)
+{
+	LOG_DBG("");
+
+	struct nrf_rpc_uart *uart_config = transport->ctx;
+
+	k_work_init(&uart_config->work, rpc_tr_uart_handler);
+
+	if (uart_config->used) {
+		return 0;
+	}
+
+	if (receive_cb == NULL) {
+		LOG_ERR("No transport receive callback");
+		return -NRF_EINVAL;
+	}
+
+	/* Initialize UART driver */
+	if (!device_is_ready(uart_config->uart)) {
+		LOG_ERR("UART device not found!");
+		return -NRF_EAGAIN;
+	}
+
+	/* Setup nRF RPC UART transport instance */
+	uart_config->receive_cb = receive_cb;
+	uart_config->context = context;
+	uart_config->used = true;
+
+	uart_irq_callback_user_data_set(uart_config->uart, serial_cb, (void *)transport);
+	uart_irq_rx_enable(uart_config->uart);
+
+	LOG_DBG("init ok");
+
+	return 0;
+}
+
+static int send(const struct nrf_rpc_tr *transport, const uint8_t *data, size_t length)
+{
+	LOG_DBG("");
+	struct nrf_rpc_uart *uart_config = transport->ctx;
+
+	if (!uart_config->used) {
+		LOG_ERR("nRF RPC transport is not initialized");
+		return -NRF_EFAULT;
+	}
+
+	LOG_DBG("Sending %u bytes", length);
+	DUMP_LIMITED_DBG(data, length, "Data: ");
+
+	/* Add UART transport header */
+	uart_poll_out(uart_config->uart, 'U');
+	uart_poll_out(uart_config->uart, 'A');
+	uart_poll_out(uart_config->uart, 'R');
+	uart_poll_out(uart_config->uart, 'T');
+	/* Add length */
+	uart_poll_out(uart_config->uart, 0xFF & length);
+	uart_poll_out(uart_config->uart, 0xFF & (length >> 8));
+	/* Add CRC (not computed for now) */
+	uart_poll_out(uart_config->uart, 0);
+
+	for (size_t i = 0; i < length; i++) {
+		uart_poll_out(uart_config->uart, data[i]);
+	}
+
+	k_free((void *)data);
+
+	LOG_DBG("exit");
+
+	return 0;
+}
+
+static void *tx_buf_alloc(const struct nrf_rpc_tr *transport, size_t *size)
+{
+	LOG_DBG("");
+	void *data = NULL;
+	struct nrf_rpc_uart *uart_config = transport->ctx;
+
+	if (!uart_config->used) {
+		LOG_ERR("nRF RPC transport is not initialized");
+		goto error;
+	}
+
+	data = k_malloc(*size);
+	if (!data) {
+		LOG_ERR("Failed to allocate Tx buffer.");
+		goto error;
+	}
+
+	return data;
+
+error:
+	/* It should fail to avoid writing to NULL buffer. */
+	k_oops();
```

Comment:
Probably it is not a good idea to crash the device when it runs out of memory.
Would it be possible to return an error to the application instead?

Reply:
yes that would be nicer. Just to note that this code is already live in a bunch of devices: I copied it from `nrf/subsys/nrf_rpc/nrf_rpc_ipc.c`.